File: android/app/src/main/java/com/modi/connect/links/bluetooth/BluetoothLink.kt
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.modi.connect.links.bluetooth

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothClass
import android.bluetooth.BluetoothDevice
import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjection
import com.modi.connect.ConnectionState
import com.modi.connect.ConnectionStateManager
import com.modi.connect.StreamingService
import com.modi.connect.audio.AudioPipeline
import com.modi.connect.core.adapters.BluetoothTransport
import com.modi.protocol.PacketHeaderCodec
import com.modi.connect.core.infrastructure.Log
import com.modi.protocol.Packet
import com.modi.connect.links.ILink
import com.modi.connect.links.LinkState
import com.modi.connect.links.LinkManager
import com.modi.connect.links.LinkParams
import com.modi.protocol.LinkType
import com.modi.protocol.PacketType
import com.modi.connect.session.DisconnectReason
import com.modi.connect.session.SessionControlMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.UUID

/**
 * BluetoothLink — 蓝牙 RFCOMM 链路（Android 端，主动发起方）
 *
 * 职责：发现已配对 Windows 设备 → 连接 RFCOMM → 发 HELLO → 推流。
 * 与 LAN 模式对称：手机点按钮主动发起连接。
 *
 * 握手方向：Android 发 HELLO(token) → Windows 校验 → 回 HELLO_ACK(route)
 * 数据通路：AudioPipeline → EncodeSender → BluetoothTransport.sendBlocking()
 */
class BluetoothLink(
    private val context: Context,
    private val pipe: AudioPipeline,
    private val stateManager: ConnectionStateManager
) : ILink {

    companion object {
        private const val TAG = "BluetoothLink"
    }

    // ── 子模块 ──
    private val connectMutex = Mutex()
    private var btTransport: BluetoothTransport? = null

    // ── ILink 状态 ──
    @Volatile override var isStreaming = false
        private set
    override val state: LinkState
        get() = if (isStreaming) LinkState.STREAMING else LinkState.IDLE
    override var onStatusChanged: ((String) -> Unit)? = null
    override var onStreamingChanged: ((Boolean) -> Unit)? = null
    @Volatile override var sessionId: UUID? = null
        private set

    @Volatile var currentRoute: Int = 0
        private set

    // ── ILink 实现 ──

    /**
     * 连接蓝牙链路（用户点击“蓝牙直连”触发）
     * 流程：查找已配对电脑 → RFCOMM 连接 → 发 HELLO 握手 → 启动推流
     */

    @SuppressLint("MissingPermission")
    override suspend fun connect(params: LinkParams): Boolean = connectMutex.withLock {
        currentRoute = params.route
        onStatusChanged?.invoke("蓝牙：搜索已配对的电脑...")
        stateManager.update(ConnectionState.CONNECTING)

        // 1. 查找已配对的 Windows 设备
        val device = findPairedWindowsDevice()
        if (device == null) {
            onStatusChanged?.invoke("蓝牙：未找到已配对的电脑（请先在系统设置中配对）")
            stateManager.update(ConnectionState.ERROR)
            return@withLock false
        }
        Log.i(TAG, "Found paired device: ${device.name} (${device.address})")
        onStatusChanged?.invoke("蓝牙：连接 ${device.name}...")

        // 2. 建立 RFCOMM 连接
        val transport = BluetoothTransport()
        btTransport = transport
        val connected = withContext(Dispatchers.IO) { transport.connectTo(device) }
        if (!connected) {
            btTransport = null
            onStatusChanged?.invoke("蓝牙：连接失败（电脑端是否已启动？）")
            stateManager.update(ConnectionState.ERROR)
            return@withLock false
        }
        onStatusChanged?.invoke("蓝牙：已连接，握手中...")

        // 3. 主动发 HELLO → 等待 ACK
        val handshake = withContext(Dispatchers.IO) {
            BtHandshakeClient.sendHelloAndWaitForAck(transport, params.route, UUID.randomUUID())
        }
        if (handshake == null) {
            onStatusChanged?.invoke("蓝牙：握手失败（电脑未响应）")
            stateManager.update(ConnectionState.ERROR)
            disconnect()
            return@withLock false
        }

        stateManager.update(ConnectionState.CONNECTED)
        sessionId = handshake.sessionId
        onStatusChanged?.invoke("蓝牙：握手成功 ✓ 准备推流")

        // 4. 启动推流（注入 BluetoothTransport）
        val capMode = LinkManager.routeToCapture(params.route)
        pipe.onFirstFrame = { stateManager.update(ConnectionState.STREAMING) }
        val ok = withContext(Dispatchers.IO) {
            pipe.currentLinkType = LinkType.BLUETOOTH
            pipe.startStreamingWithTransport(transport, capMode, params.proj, context)
        }

        if (ok) {
            isStreaming = true
            onStatusChanged?.invoke("蓝牙推流中：路线${params.route + 1}")
            onStreamingChanged?.invoke(true)
            context.startForegroundService(Intent(context, StreamingService::class.java))
        } else {
            onStatusChanged?.invoke("蓝牙：启动推流失败")
            stateManager.update(ConnectionState.ERROR)
            disconnect()
        }
        ok
    }

    /** 推流中热切路线：切换采集模式 + 发送 ROUTE 包通知 Windows 切换 AudioRouter */
    override suspend fun sendRouteUpdate(route: Int, proj: MediaProjection?): Boolean {
        if (!isStreaming) { currentRoute = route; return true }
        currentRoute = route

        val capMode = LinkManager.routeToCapture(route)
        val ok = withContext(Dispatchers.IO) { pipe.switchMode(capMode, proj, context) }
        if (!ok) { onStatusChanged?.invoke("需先授权系统音频"); return false }

        // 发送 ROUTE 包到 Windows
        val transport = btTransport ?: return false
        withContext(Dispatchers.IO) {
            val protocol = PacketHeaderCodec()
            val payload = byteArrayOf(route.toByte())
            val packet = Packet(PacketType.ROUTE, LinkType.BLUETOOTH, 0u, payload)
            transport.sendBlocking(protocol.encode(packet))
        }
        return true
    }

    override suspend fun sendDisconnectRequest(targetLink: Byte, reason: DisconnectReason): Boolean {
        val currentId = sessionId ?: return false
        val transport = btTransport?.takeIf { it.isConnected } ?: return false
        val message = SessionControlMessage.request(currentId, LinkType.BLUETOOTH, targetLink, reason)
        return withContext(Dispatchers.IO) {
            val encoded = PacketHeaderCodec().encode(message.toPacket())
            transport.sendBlocking(encoded)
            transport.isConnected
        }
    }

    /** 断开蓝牙链路：停止推流 + 关闭 RFCOMM + 状态回退 */
    override fun disconnect() {
        context.stopService(Intent(context, StreamingService::class.java))
        pipe.stopStreaming()

        btTransport?.let { t ->
            kotlinx.coroutines.runBlocking { t.disconnect() }
        }
        btTransport = null
        sessionId = null

        isStreaming = false
        stateManager.update(ConnectionState.DISCONNECTED)
        onStatusChanged?.invoke("蓝牙：已停止")
        onStreamingChanged?.invoke(false)
    }

    // ── 发现已配对设备（按蓝牙设备类过滤，只连电脑类设备） ──

    @SuppressLint("MissingPermission")
    private fun findPairedWindowsDevice(): BluetoothDevice? {
        val adapter = BluetoothAdapter.getDefaultAdapter() ?: return null
        if (!adapter.isEnabled) return null

        val bondedDevices = adapter.bondedDevices ?: return null

        // 优先找电脑类设备（Major Class = COMPUTER）
        val computer = bondedDevices.firstOrNull { device ->
            device.bluetoothClass?.majorDeviceClass == BluetoothClass.Device.Major.COMPUTER
        }
        if (computer != null) return computer

        // 找不到电脑类设备时返回 null（不连耳机/音箱）
        Log.w(TAG, "No COMPUTER-class device found in ${bondedDevices.size} paired devices")
        return null
    }

}

File: windows/MoDi.Desktop/Links/Bluetooth/BluetoothLink.cs
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
using System;
using System.Threading;
using System.Threading.Tasks;
using MoDi.Core;
using MoDi.Protocol;
using MoDi.Core.Adapters;
using MoDi.Core.Factory;
using MoDi.Core.Infrastructure;
using MoDi.Desktop.Core.Session;

namespace MoDi.Desktop.Links;

/// <summary>
/// BluetoothLink — 蓝牙 RFCOMM 链路（常驻服务）
///
/// 职责：常驻监听 RFCOMM → 等待 Android 连接 → 被动接收 HELLO 回 ACK → AudioEngine 播放。
/// 与 WifiLanLink 对称：开机即启动，始终等待手机连接。
///
/// 数据通路：BluetoothTransport.PacketReceived → AudioEngine（Opus 解码 → 播放）
/// 握手方向：Android 发 HELLO(token) → Windows 校验 → 回 HELLO_ACK(route)（与 LAN 一致）
/// </summary>
public sealed class BluetoothLink : ILink
{
    private const string Tag = "BluetoothLink";

    // ── 核心模块 ──
    private BluetoothTransport? _transport;
    private AudioEngine? _engine;

    /// <summary>当前活跃的 AudioEngine（会话期间非 null，LinkManager 用于 Volume 控制）</summary>
    public AudioEngine? ActiveEngine => _engine;
    private readonly ConnectionStateManager _stateManager;
    private CancellationTokenSource? _cts;
    private Task? _listenLoop;
    private volatile bool _started;

    // ── 事件（LinkManager / UI 订阅） ──
    public Action<string>? OnStatusChanged;
    public Action<bool>? OnActiveChanged;
    public Action<int>? OnRouteChanged;
    /// <summary>蓝牙会话开始（手机连接+握手成功），LinkManager 用于暂停 LAN 引擎</summary>
    public Action<Guid>? OnSessionStarted;
    /// <summary>蓝牙会话结束（手机断开），LinkManager 用于恢复 LAN 引擎</summary>
    public Action<Guid>? OnSessionEnded;
    public Func<SessionControlMessage, byte, SessionControlMessage>? OnDisconnectRequest;

    public LinkState State { get; private set; } = LinkState.Idle;
    public bool IsActive => _started;

    public BluetoothLink(ConnectionStateManager stateManager)
    {
        _stateManager = stateManager;
    }

    // ── ILink 实现 ──

    /// <summary>启动蓝牙链路（常驻监听，与 LAN 一起开机启动）</summary>
    public Task<bool> ConnectAsync()
    {
        if (_started) return Task.FromResult(true);
        _started = true;
        State = LinkState.Listening;

        _cts = new CancellationTokenSource();
        var ct = _cts.Token;

        // 启动常驻监听循环（后台）
        _listenLoop = Task.Run(() => ListenLoopAsync(ct));

        OnStatusChanged?.Invoke("蓝牙：就绪，等待手机连接");
        OnActiveChanged?.Invoke(true);
        Log.I(Tag, "Bluetooth link started (resident)");
        return Task.FromResult(true);
    }

    /// <summary>停止蓝牙链路（关闭监听 + 断开当前连接）</summary>
    public async Task DisconnectAsync()
    {
        if (!_started) return;
        _started = false;
        State = LinkState.Idle;

        _cts?.Cancel();
        _cts?.Dispose();
        _cts = null;

        if (_listenLoop != null)
        {
            try { await _listenLoop; } catch { }
            _listenLoop = null;
        }

        await CleanupSessionAsync();
        _transport?.StopListening();
        _transport?.Dispose();
        _transport = null;

        OnStatusChanged?.Invoke("蓝牙：已停止");
        OnActiveChanged?.Invoke(false);
    }

    // ── 常驻监听循环（核心状态机） ──
    // 流程：等待连接 → 被动握手 → 创建引擎播放 → 等待断开 → 清理 → 循环

    private async Task ListenLoopAsync(CancellationToken ct)
    {
        _transport = new BluetoothTransport();
        _transport.StartListening();

        while (!ct.IsCancellationRequested)
        {
            try
            {
                // 等待 Android 连接
                OnStatusChanged?.Invoke("蓝牙：等待手机连接...");
                var connected = await _transport.WaitForConnectionAsync(ct);
                if (!connected) continue;

                // 连接建立 → 等待 HELLO
                OnStatusChanged?.Invoke("蓝牙：手机已连接，等待握手...");
                var handshake = await BtPassiveHandshake.WaitForHelloAsync(_transport, ct);
                if (!handshake.HasValue)
                {
                    OnStatusChanged?.Invoke("蓝牙：握手失败，重新等待...");
                    await _transport.DisconnectAsync();
                    continue;
                }

                // 握手成功 → 创建 AudioEngine → 播放
                _stateManager.BeginConnecting();
                var sessionId = handshake.Value.SessionId;
                OnSessionStarted?.Invoke(sessionId);
                StartAudioEngine(handshake.Value.Route);
                OnRouteChanged?.Invoke(handshake.Value.Route);
                OnStatusChanged?.Invoke($"蓝牙：推流中 ✓ 路线{handshake.Value.Route + 1}");

                // 监听 ROUTE 热切包（AudioEngine 只处理 Audio 包，ROUTE 由此处理）
                _transport.PacketReceived += OnSessionPacket;

                // 等待连接断开（ReadLoop 退出 = 连接断开）
                await WaitForDisconnectAsync(ct);

                // 断开 → 清理 → 重新等待
                _transport.PacketReceived -= OnSessionPacket;
                OnSessionEnded?.Invoke(sessionId);
                OnStatusChanged?.Invoke("蓝牙：手机断开，重新等待...");
                CleanupAudioEngine();
                await _transport.DisconnectAsync();
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                Log.E(Tag, $"ListenLoop error: {ex.Message}");
                if (!ct.IsCancellationRequested)
                    await Task.Delay(2000, ct);  // 防止快速循环
            }
        }
    }

    // ── ROUTE 热切（命名方法，确保可取消订阅） ──
    // 推流中手机切换路线时，发送 ROUTE 包，此处解码并切换 AudioRouter 模式

    private void OnSessionPacket(ReadOnlyMemory<byte> data)
    {
        var protocol = PlatformFactory.CreateProtocol();
        var decoded = protocol.Decode(data.Span);
        if (!decoded.HasValue) return;
        if (decoded.Value.Type == PacketType.Route)
        {
            BtPassiveHandshake.HandleRoutePacket(data, _engine, OnStatusChanged, OnRouteChanged);
            return;
        }
        if (decoded.Value.Type != PacketType.Data ||
            !SessionControlMessage.TryDecode(decoded.Value.Payload, out var request) ||
            request.Action != SessionControlAction.DisconnectRequest ||
            OnDisconnectRequest is not { } handleDisconnect)
            return;

        var ack = handleDisconnect(request, LinkType.Bluetooth);
        _ = AcknowledgeThenCloseAsync(protocol, ack);
    }

    private async Task AcknowledgeThenCloseAsync(IPacketProtocol protocol, SessionControlMessage ack)
    {
        var transport = _transport;
        if (transport == null) return;
        await transport.SendAsync(protocol.Encode(ack.ToPacket()));
        if (ack.Result == DisconnectResult.Accepted)
            await transport.DisconnectAsync();
    }

    internal void StopCurrentSession(bool closeTransport)
    {
        CleanupAudioEngine();
        if (closeTransport && _transport != null)
            _ = _transport.DisconnectAsync();
    }

    // ── AudioEngine 管理（蓝牙专属引擎，不复用 LAN 引擎） ──

    /// <summary>创建蓝牙专属 AudioEngine，设置路由模式并启动播放</summary>

    private void StartAudioEngine(int route)
    {
        var speaker = PlatformFactory.CreateRenderer(useCable: false);
        var cable = PlatformFactory.CreateRenderer(useCable: true);
        _engine = new AudioEngine(_transport, speaker, cable);
        _engine.Router.SetMode(BtPassiveHandshake.RouteToMode(route));

        _engine.OnFirstFrameDecoded += () => _stateManager.Update(ConnectionState.Streaming);
        _engine.Start();
        _stateManager.Update(ConnectionState.Connected);
    }

    /// <summary>停止并释放 AudioEngine</summary>
    private void CleanupAudioEngine()
    {
        _engine?.Stop();
        _engine?.Dispose();
        _engine = null;
    }

    /// <summary>轮询等待 RFCOMM 连接断开（ReadLoop 退出后 IsConnected 变 false）</summary>
    private async Task WaitForDisconnectAsync(CancellationToken ct)
    {
        // 等待 transport 连接断开（IsConnected 变 false）
        while (!ct.IsCancellationRequested && _transport?.IsConnected == true)
            await Task.Delay(500, ct);
    }

    /// <summary>清理当前会话（引擎 + 传输层）</summary>
    private async Task CleanupSessionAsync()
    {
        CleanupAudioEngine();
        if (_transport != null)
            await _transport.DisconnectAsync();
    }

    public void Dispose()
    {
        _cts?.Cancel();
        _cts?.Dispose();
        _engine?.Dispose();
        _transport?.Dispose();
    }
}
